src/strings/suf_ary.rs

```diff
@@ -0,0 +1,151 @@
+//! # Suffix Array
+
+use std::cmp::Ordering;
+use std::ops::Range;
+
+use crate::data_structures::rmq::RMQ;
+use ac_library::string::{lcp_array_arbitrary, suffix_array_manual};
+
+/// # Example
+/// ```
+/// use programming_team_code_rust::helpers::compress::compress;
+/// use programming_team_code_rust::strings::suf_ary::SufAry;
+///
+/// let s = "banana";
+/// let suf_ary1 = SufAry::new(&s.chars().map(|c| c as usize).collect::<Vec<usize>>(), 255);
+///
+/// // 0 banana 3
+/// // 1 anana  2
+/// // 2 nana   5
+/// // 3 ana    1
+/// // 4 na     4
+/// // 5 a      0
+/// //
+/// // 5 a      0
+/// //   |
+/// // 3 ana    1
+/// //   |||
+/// // 1 anana  2
+/// //
+/// // 0 banana 3
+/// //
+/// // 4 na     4
+/// //   ||
+/// // 2 nana   5
+///
+/// assert_eq!(suf_ary1.sa, [5, 3, 1, 0, 4, 2]);
+/// assert_eq!(suf_ary1.sa_inv, [3, 2, 5, 1, 4, 0]);
+/// assert_eq!(suf_ary1.lcp, [1, 3, 0, 0, 2]);
+///
+/// let a = [-4, 8, 1_000_000_000, 3];
+/// let (a_comp, max_val) = compress(&a);
+/// let suf_ary2 = SufAry::new(&a_comp, max_val);
+///
+/// assert_eq!(suf_ary2.sa, [0, 3, 1, 2]);
+/// ```
+pub struct SufAry {
+    n: usize,
+    s: Vec<usize>,
+    /// suffix array
+    pub sa: Vec<usize>,
+    /// inverse suffix array
+    pub sa_inv: Vec<usize>,
+    /// longest common prefix array
+    pub lcp: Vec<usize>,
+    rmq: RMQ<usize, fn(usize, usize) -> usize>,
+}
+
+impl SufAry {
+    /// Creates a new Suffix Array struct
+    ///
+    /// # Complexity
+    /// - Time: O(n + max_val)
+    /// - Space: O(n)
+    pub fn new(s: &[usize], max_val: usize) -> Self {
+        let sa = suffix_array_manual(
```

Comment:
IK we previously discussed multiple constructors (one taking string as [T], and the other taking string as [usize] or [char] maybe); but I decided against it:

basically we need to store the string as a member var. So in order to have multiple constructors with different types, we need to templatize the string type; which made it less clean for similar reasons why templatized binary trie was less clean (you need a template type with a ton of traits)

So instead I opted for a single constructor, taking [usize]. If the input is a string, you can just map it to this type.

If the input is an array, you can call the `compress` function first

Reply:
sounds good

Reply:
yeah the compress as a separate helper makes a lot of sense, it really doesn't belong to the SA